### ISSUE TITLE
Atom gets really slow after a certain time using markdown-preview

### ISSUE THREAD

**Issue by Phisherman** [atom/atom#6303]
I usually use ATOM to write markdown and also use the markdown-preview which works beautiful. But after a certain time (or is it caused by saving multiple times and editing texts?) I often experience that the programme stucks. After restarting it works fine.

**Comment by izuzak** [comment]
@Phisherman Thanks for taking the time to share this feedback, but I'm wondering if you could provide more details? Specifically:
* which version of Atom and which OS are you using?
* what does "really slow" mean -- how does this slowness manifest?
* what does "certain time" mean -- is that typically 1 hour or 1 day or 1 week?
* what does "programme stucks" mean -- does Atom stop reacting to your input completely or is it just very slow?
* do these problems happen consistently for you? If so, can you share more information about what files your editing or even post those files somewhere and provide a link?
* do you have any packages installed and do you see any improvement if you run Atom in [safe mode](https://atom.io/docs/latest/hacking-atom-debugging#check-if-the-problem-shows-up-in-safe-mode)? `apm list --installed` gives you a list of installed packages

**Comment by Phisherman** [comment]
Sorry for my unclear specification!

- which version of Atom and which OS are you using?

Atom is version 0.190.0 and OS is Win 8.1 Pro 64bit.

- what does "really slow" mean -- how does this slowness manifest? what does "programme stucks" mean -- does Atom stop reacting to your input completely or is it just very slow?

The programme starts to stuck. When I write sth. the text cursor starts to lag behind. After 10 to 15min the programme gets more and more "stucky" until it hangs up to 5-10s. When the text cursor lags behind the UI freezes also.

- what does "certain time" mean -- is that typically 1 hour or 1 day or 1 week?
I can't determinate it exactly because I had Atom open for at least 3-4h but didn't use it the whole time.

- do these problems happen consistently for you? If so, can you share more information about what files your editing or even post those files somewhere and provide a link?

I use it just for editing markdown but heard of colleagues (e.g. @squarerootfury ) that they experienced exact the same problems i described.

- do you have any packages installed and do you see any improvement if you run Atom in safe mode? apm list --installed gives you a list of installed packages

I have no packages installed.

**Comment by izuzak** [comment]
Thanks for the details, @Phisherman. I haven't observed this before, so I'd like to ask for more information again. Could you perhaps share one such file which you were editing and for which you observed this? I'd love to try and reproduce this problem.

**Comment by st3w4r** [comment]
Hi, i have the same problem, when i open a Markdown Preview editor, Atom become very slow and then end up crashed.

Atom version is 0.190.0 on Mac OS X Yosemite 10.10.3 (14D131)

Here is the Crash Log:
[Crash Log](http://pastebin.com/mZX2Pd5y)

**Comment by surdu** [comment]
I have the same issue as mentioned above. What I have noticed is that starting atom in safe mode will make the problem not appear anymore.

The only thing that safe mode does is disable all the community packages ?

**Comment by mhnavarro** [comment]
I got this problem too. Windows 8.1 x64, atom version 0.201.0

I checked the task manager and....

![github](https://cloud.githubusercontent.com/assets/7016419/7789667/2496a022-0226-11e5-93a1-9cae4c172efe.png)

This is what I get from apm list:
´´´
Built-in Atom packages (88)
├── archive-view@0.57.0
├── atom-dark-syntax@0.27.0
├── atom-dark-ui@0.49.0
├── atom-light-syntax@0.28.0
├── atom-light-ui@0.41.0
├── autocomplete-atom-api@0.9.0
├── autocomplete-css@0.7.2
├── autocomplete-html@0.7.2
├── autocomplete-plus@2.17.1
├── autocomplete-snippets@1.6.1
├── autoflow@0.24.0
├── autosave@0.20.0
├── background-tips@0.24.0
├── base16-tomorrow-dark-theme@0.26.0
├── base16-tomorrow-light-theme@0.9.0
├── bookmarks@0.35.0
├── bracket-matcher@0.74.0
├── command-palette@0.36.0
├── deprecation-cop@0.51.0
├── dev-live-reload@0.46.0
├── encoding-selector@0.20.0
├── exception-reporting@0.24.0
├── find-and-replace@0.167.0
├── fuzzy-finder@0.87.0
├── git-diff@0.55.0
├── go-to-line@0.30.0
├── grammar-selector@0.47.0
├── image-view@0.54.0
├── incompatible-packages@0.24.0
├── keybinding-resolver@0.33.0
├── language-c@0.44.0
├── language-clojure@0.15.0
├── language-coffee-script@0.40.0
├── language-csharp@0.5.0
├── language-css@0.29.0
├── language-gfm@0.76.0
├── language-git@0.10.0
├── language-go@0.26.0
├── language-html@0.37.0
├── language-hyperlink@0.13.0
├── language-java@0.15.0
├── language-javascript@0.77.0
├── language-json@0.14.0
├── language-less@0.27.0
├── language-make@0.14.0
├── language-mustache@0.11.0
├── language-objective-c@0.15.0
├── language-perl@0.24.0
├── language-php@0.24.0
├── language-property-list@0.8.0
├── language-python@0.35.0
├── language-ruby@0.54.0
├── language-ruby-on-rails@0.21.0
├── language-sass@0.38.0
├── language-shellscript@0.14.0
├── language-source@0.9.0
├── language-sql@0.15.0
├── language-text@0.6.0
├── language-todo@0.21.0
├── language-toml@0.16.0
├── language-xml@0.30.0
├── language-yaml@0.22.0
├── link@0.30.0
├── markdown-preview@0.149.0
├── metrics@0.51.0
├── notifications@0.50.0
├── one-dark-syntax@0.5.0
├── one-dark-ui@0.8.2
├── one-light-syntax@0.6.0
├── one-light-ui@0.8.2
├── open-on-github@0.37.0
├── package-generator@0.39.0
├── release-notes@0.52.0
├── settings-view@0.205.0
├── snippets@0.90.0
├── solarized-dark-syntax@0.35.0
├── solarized-light-syntax@0.21.0
├── spell-check@0.58.0
├── status-bar@0.72.0
├── styleguide@0.44.0
├── symbols-view@0.97.0
├── tabs@0.68.0
├── timecop@0.31.0
├── tree-view@0.171.0
├── update-package-dependencies@0.10.0
├── welcome@0.27.0
├── whitespace@0.29.0
└── wrap-guide@0.33.0

C:\Users\Marco\.atom\packages (15)
├── atom-html-preview@0.1.6
├── atom-terminal-panel@4.4.4
├── autoclose-html@0.16.2
├── autocomplete-php@0.3.6
├── javascript-snippets@1.0.0
├── language-aspx@0.4.0
├── language-cshtml@0.1.1
├── language-latex@0.6.1
├── markdown-toc@0.2.1
├── pdf-view@0.22.0
├── preview-plus@1.1.42
├── steam-pirate-syntax@0.3.0
├── steam-pirate-ui@0.4.0
├── texlicious@0.4.2
└── wordcount@2.2.2

´´´

**Comment by Nijikokun** [comment]
I've had this happen, disabled a language syntax that was very heavy compared to the default (`language-babel` compared to `language-javascript`) and the input went back to normal.

I've also had it happen during linting or spell-check.

**Comment by bentolor** [comment]
I was just to about to open a new issue, but stumbled over this one: **I can fully confirm this issue!** The editor window of Atom gets _sluggish_ after a certain time of usage in a markdown document (500 lines total).

The effect: Keystrokes start to appear only after a latency of 150ms up to multiple seconds. I just did a quick CPU profiling while only entering/deleting bogus letters. Delays up to multiple seconds. Please see the attached CPU screens.

Using the Markdown Preview makes this worse due large amounts of garbage collections.

I already disabled the autocompletion, which was also causing large latencies: https://github.com/atom/autocomplete-plus/issues/501

How can we help to provide the missing informations?

![cpu-chart](https://cloud.githubusercontent.com/assets/588260/8108891/aa5d1200-1053-11e5-8e67-852e8c64329b.png)
![cpu-heavy](https://cloud.githubusercontent.com/assets/588260/8108893/acc6bd98-1053-11e5-8370-09126f9a64cf.png)

**Comment by bentolor** [comment]
Here another results: Only after a few minutes editing has become unbearable. CPU is 73% garbage collection
![gc](https://cloud.githubusercontent.com/assets/588260/8110705/174ee15e-105e-11e5-9a0e-4746ce64d61b.png)
![unusable-chart](https://cloud.githubusercontent.com/assets/588260/8110709/1a768ada-105e-11e5-9927-b6d2fa9bc9be.png)

**Comment by mnquintana** [comment]
@bentolor It would be extremely helpful if you could reproduce this in safe mode (`atom --safe`).

**Comment by bentolor** [comment]
@mnquintana `--help` says: about `--safe`: _Do not load packages from ~/.atom/packages or ~/.atom/dev/packages_. Both are empty. Does it anything more?

**Comment by 50Wliu** [comment]
It also doesn't load your `init.coffee`, `keymap.cson`, and `config.cson` files.

**Comment by joshfester** [comment]
I get this behavior as well. After a while, my PC becomes so sluggish that the only solution is to reboot. The time varies but sometimes it only takes like 20 minutes. It reaches a point where nothing is usable. Atom  0.211.0  on Ubuntu 15.04

**Comment by joshfester** [comment]
I get this behavior as well. After a while, my PC becomes so sluggish that the only solution is to reboot. The time varies but sometimes it only takes like 20 minutes. It reaches a point where nothing is usable. Atom 0.211.0 on Ubuntu 15.04

**Comment by chrisvfritz** [comment]
Can confirm this on OS X, 2014 MacBook Air. I didn't try `atom --safe`, because whether the problem is caused by external packages or not, Atom is not very useful to me without those packages. It does appear to happen across file types. I'd be happy to share more info on packages installed or anything else.

**Comment by MichaelAquilina** [comment]
I've noticed similar issues. It also *seems* to be worse if you have multiple panes open.

**Comment by devfoo-one** [comment]
:+1: same here with Atom 1.0.2 on OS X 10.9.5

**Comment by joshfester** [comment]
This got so bad for me that I had to give up and switch to Brackets ;(

**Comment by ShiftyMcCool** [comment]
I'm a Sublime user but I like the philosophy behind Atom (plus I'm a Git-nut).  I started a new project with Atom and while editing a file that was only 400 lines, the lag from typing and deleting was so much that I had to abandon Atom and switch back to Sublime.  I'll add as many details as I can, let me know if more would be helpful:
OS: Ubuntu 14.10
Hardware: Not an issue... trust me.  I'll supply more if needed, but it's a beast.
File edited was Python
No custom theming, stayed with stock theme.
apm list:
In addition to the 88 built-ins:
/home/rharris/.atom/packages (9)
├── autocomplete-python@0.5.1
├── file-icons@1.5.8
├── git-plus@5.2.2
├── highlight-selected@0.10.1
├── minimap@4.12.0
├── minimap-highlight-selected@4.3.0
├── project-manager@1.15.11
├── python-nosetests@0.2.0
└── todo-show@0.13.0

**Comment by mnquintana** [comment]
@ShiftyMcCool Do you get the same lag / slow-down in safe mode? (`atom --safe`)

**Comment by benogle** [comment]
Based on everyone's reports this could be a few things

* The undo stack is getting huuuuge
* There is a leak in markdown preview
* There is a leak somewhere else

So some questions

Are you working on the same file the entire time when it becomes sluggish? When you close all the files/tabs and wait for a few seconds (say 30), does it go reasonably back to normal?

Are all of you working on markdown documents with markdown preview open?

Thanks @bentolor for the screenshots. Can more of you with this issue take some screenshots of a profile when it gets slow? The most helpful view would be the bottom up list. Flame graphs are useful when the function names are visible. I just want to see if we are converging on a similar problem from everyone or if we're looking at lots of different issues.

**Comment by maxbrunsfeld** [comment]
@bentolor It looks like the spell-checker is creating lots of markers which are being snapshotted before and after every change.

What version of Atom are you on? I thought that I fixed this several versions back.

**Comment by maxbrunsfeld** [comment]
Also, I'm really sorry that you're experiencing that kind of lag. That's really crappy.

**Comment by MichaelAquilina** [comment]
I am personally not working with markdown preview, just python files. Closing atom and re-opening it does not seem to solve the issue. In my case I have to switch to one pane mode to make atom usable again. Does atom run in the background when closed? Maybe I should try kill -9 it to be sure I have a fresh state the next time.

**Comment by surdu** [comment]
@mnquintana Now I'm not experiencing this issue anymore, but when I did it was not reproducing in safe mode, only in "normal" mode.

**Comment by bentolor** [comment]
@benogle @maxbrunsfeld  Thank you for looking into this issue!

- At time of writing my my markdown document I was using a vanilla Atom 0.208.0. 
- I was working on a single markdown document (article) which contained mostly foreign words
- It appeared to me, that the preview accelerates the effect. As the document became larger, I only used the preview on demand. 
- It did feel like a polluted heap with permanent GCs in the background.
- I tried to create a «steps to replicate» in `--safe` mode but failed: **In my case it occurs only after longer session of real-life work involving all sorts of edits.** 

My guess would be that the root issue is rather located in edit/undo or spell checking. Currently I'm no longer editing larger markdown documents and the effect only occurred on doing *real work* so I don't have any feedback for 1.0.2 for now.

**Comment by ShiftyMcCool** [comment]
@mnquintana I just tried with --safe and it seems worse, if that's possible.  I'm working in the same file, but i have added quite a bit of text to that file from sublime.  Characters show up several seconds after the keystrokes and mouse selection is sluggish but not nearly as severe as typing.  Let me know if you need anything else, I'll do what I can.

**Comment by benogle** [comment]
@ShiftyMcCool saving a profile or timeline would be really helpful

**Comment by ShiftyMcCool** [comment]
@benogle I'm not sure what you mean.  If you let me know what info you're interested in I'll see what I can do.

**Comment by benogle** [comment]
@ShiftyMcCool See https://atom.io/docs/latest/hacking-atom-debugging#diagnose-performance-problems-with-the-dev-tools-cpu-profiler

**Comment by ShiftyMcCool** [comment]
Looks like the main culprit is WindowEventHandler.prototype.onKeydown
I tried saving the profile, but the file keeps ending up empty.  Here's a screenshot:

![atom](https://cloud.githubusercontent.com/assets/1320624/8913205/6a2c249c-344c-11e5-84f2-c0cbe5105d73.png)

**Comment by benogle** [comment]
@ShiftyMcCool Can you zoom in (scroll) on the big chunk here: 

![profile](https://cloud.githubusercontent.com/assets/69169/8913275/e797c792-344c-11e5-9cff-525258d78427.jpg)

The blocks at the bottom of the spike will be where the time is actually spent. Something is happening in the keydown event, but it's not the keydown event itself. When you zoom in, we'll need to see the names of the functions. Sad that saving produces an empty result.

**Comment by ShiftyMcCool** [comment]
Yeah, this would be a lot easier for everybody if the profile would save.  I had to re-profile so this is a different chart.  I zoomed in on a very small section to get the bottom blocks:

![atom2](https://cloud.githubusercontent.com/assets/1320624/8914325/73a447dc-3453-11e5-8837-e094d81d2ce0.png)

**Comment by benogle** [comment]
Can you also take a screenshot of the `Heavy: bottom up` option?

**Comment by maxbrunsfeld** [comment]
@ShiftyMcCool In this screenshot, it looks like a lot of time is spent in the [`minimap-highlight-selected`](https://github.com/atom-minimap/minimap-highlight-selected/blob/master/lib/minimap-highlight-selected-view.coffee#L10) package. Do things still slow down if you disable that package?

**Comment by Allardvm** [comment]
I've been experiencing very similar issues for the past days, where typing feels much slower and is quite laggy. I'm not sure if it's because my code is longer (~1000 loc), or because of Atom-related changes (although I did not install any new plugins and the problem also occurs without non-native plugins). Please find attached a zoomed `chart` and `Heavy (Bottom Up)` table of a few seconds worth of profiling while I type. Most time seems to be spend in the `Node.getEnd` function in `marker-index.js:186`.

![screen shot 2015-09-11 at 17 20 37](https://cloud.githubusercontent.com/assets/11735900/9818690/249c1d92-58aa-11e5-862d-b5f10ae84e6d.png)

![screen shot 2015-09-11 at 17 22 04](https://cloud.githubusercontent.com/assets/11735900/9818691/25c29976-58aa-11e5-8b8e-2f3b3d2cce51.png)

If there's anything I can do to help, please let me know.

**Comment by Allardvm** [comment]
For your information: I managed to solve the issue by deleting the files in `.atom/storage`. There were several relatively sizable files in there (~1mb), one of which was belonged to the code I was working on. After the delete, the new version of this file was much smaller and typing was back to normal. The new profile showed that almost no time was spend on `Node.getEnd` and all of the other functions that are above `set.value` in the log in the profile log above this post. My suspicion is that the `undo` history had grown huge, which might be related to 2016295dcd32d89d804e9b3321c90c3f875a64c8.

**Comment by SteveViss** [comment]
In my case, the issue was related to the [pdf-view package](https://atom.io/packages/pdf-view). I just removed it and the CPU performance increased.

**Comment by stefankroes** [comment]
For me Atom started performing properly again after uninstalling the Google Material UI theme.

**Comment by duranmla** [comment]
:+1:

**Comment by andersennl** [comment]
Is there any update on this? I've removed the editor- files even if they weren't big on my machine but it didn't solve anything. 

It's really weird behavior, sometimes Atom runs really smooth, sometimes I have to wait several seconds for my written text to appear. I had to switch back to sublime to be able to work.

**Comment by benogle** [comment]
Random lagging could be a million things. Where _specifically_ is it laggy? Typing? Tab switching? Can you try starting atom in [safe mode](https://atom.io/docs/latest/hacking-atom-debugging#check-if-the-problem-shows-up-in-safe-mode)? Does that fix it?

**Comment by andersennl** [comment]
@benogle It's laggy while typing, but not always. It's more like when I write faster, the lag gets bigger. So it's not always there. When I begin to type much text fast, Atom will like completely freeze and then, after several seconds, suddenly start typing the characters I've entered. This happens especially when I duplicate a long, complex (maybe not related) line. (I was working on a Atom package, so I was working with CoffeeScript, if that helps.) Opening the preferences pane at the time Atom is lagging or completely frozen behaves the same as typing, the preferences won't show up the next couple of seconds.

I've already tried to deactivate the usual suspects (at least the most obvious for me), like everything related to autocompletion but that wasn't it.

Starting Atom in safe mode seems to completely prevent any lag. I will continue to deactivate packages in normal mode now, one by one, and then give an update.

**Comment by caiobep** [comment]
Hello, Im having the same issue, atom gets slow every time I try to write anything. I already tried to use atom in safe mode, but i keep having the issue. I'm running atom on OS X 10.11.2 . Is there anyway I can help you more? If yes please let me know

**Comment by lee-dohm** [comment]
@caiobep Please see [the Debugging section of the Flight Manual](https://atom.io/docs/latest/hacking-atom-debugging#diagnose-performance-problems-with-the-dev-tools-cpu-profiler) for instructions on how to use the built-in CPU profiler to track what is causing things to be slow. While the pictures (like posted above) are nice, saving the raw data in a `*.cpuprofile` file and making it available for people to look through would be the most helpful.

**Comment by caiobep** [comment]
@lee-dohm Thank you very much for the support. I've tried to record the Cpu Profile in this file:
[CPU-20151217T105338.cpuprofile.zip](https://github.com/atom/atom/files/65392/CPU-20151217T105338.cpuprofile.zip)

**Comment by tzapu** [comment]
same lagging happens to me when using markdown preview, a ~260 lines document
cpu profile
the slugginess is unbearable

cheers

[CPU-20160113T112634.cpuprofile.zip](https://github.com/atom/atom/files/88621/CPU-20160113T112634.cpuprofile.zip)

**Comment by nwinkler** [comment]
I had the same issue, and with the help of the excellent [Dev Tools link](https://atom.io/docs/latest/hacking-atom-debugging#diagnose-performance-problems-with-the-dev-tools-cpu-profiler) provided by @lee-dohm, I was able to identify the [color-gutter package](https://github.com/bkbooth/color-gutter) as the culprit. With growing file size (of a markdown file; Markdown Preview not open), it was consuming about 50% of the CPU.

After uninstalling color-gutter, Atom is back to regular speed, a lot more fun to use.

**Comment by Riveascore** [comment]
@Allardvm, how do I limit the **undo history size**?

**Comment by Allardvm** [comment]
@Riveascore I do not know how to (automatically) limit it, but I managed to resolve my slowdown by deleting the file that, as far as I can see, contains the history. You will find one file (named 'editor-[...]') for each Atom document in `.atom/storage`. After deletion, the files are automatically regenerated by Atom.

**Comment by lczech** [comment]
Having the same issue, also with markdown files, but the profile seems to be different from the ones previously posted here:

![atom_slow_md](https://cloud.githubusercontent.com/assets/4741887/13400363/a4c51724-df07-11e5-9c95-1d69a7ac67df.png)

As this shows, atom freezes for 30sec straight, which happens about every two minutes when writing a markdown doc, rendering it completely unusable. All input in this time is still processed afterwards, so you can just keep on typing, and eventually, it will catch up and display the text. Still, this is not usable.

I also tried to deactivate Markdown preview and spell checking, with no effect (as one would expect, after seeing the profile).

Using:
atom 1.5.3
ubuntu 14.04

**Comment by secretgspot** [comment]
after using for a day and having issues of random lag here and there, this obviously not ready for prime time.  Going back to Sublime Text

**Comment by codercms** [comment]
I have tried Atom 1.9.0 beta and 1.8.0 stable, but they're lagging. It's freezing while typing text in API Workbench (RAML).
Windows 10 x64, 8 GB DDR4 RAM, Intel Core i7 6700HQ.
Is there any way to fix freezing?

**Comment by DavidAwad** [comment]
+1 on this issue, it's actually ridiculously slow, on really fast machines. Not sure what's happening. It's almost unusable

**Comment by julianonunes** [comment]
I'm editing a 330 lines markdown document and also get the lag problem. Any updates on this?

**Comment by nwinkler** [comment]
@julianonunes Did you try the same thing with Atom in safe mode (i.e. without any user packages)?

See here for how to run Atom in `safe` mode: https://discuss.atom.io/t/what-does-safe-mode-do/22229

If this problem is gone while running in safe mode, then the slowness is caused by a package that you installed. In my case, it was caused by a package (_color-gutter_ to be correct). Uninstalling that package fixed the issue for me.

**Comment by julianonunes** [comment]
@nwinkler Running in safe mode it is faster but still lags a little

**Comment by Phisherman** [comment]
@all: This issue was created to find a solution for the *slowness problem with markdown-preview* in particular. I think it's running out of control because I generalized the issue title. Therefore I just redescribed it.

Don't you think it would be better to create separate issues for every problem? Iny my humble opinion this shouldn't be a collective issue.

**Comment by lee-dohm** [comment]
If this is specifically an issue with markdown-preview, we should close it and reopen it on the https://github.com/atom/markdown-preview repository.

**Comment by tzapu** [comment]
hi, there s already an issue there that might be relevant https://github.com/atom/markdown-preview/issues/197

**Comment by lee-dohm** [comment]
Thanks @tzapu. I'm going to mark this as a duplicate of atom/markdown-preview#197. People should subscribe there for updates.
